src/react/utils.js

```diff
@@ -220,19 +220,54 @@ export function getUniqueReactElementKey(index?: string, usedReactElementKeys: S
 }
 
 // a helper function to loop over ArrayValues
-export function forEachArrayValue(realm: Realm, array: ObjectValue, mapFunc: Function): void {
+export function forEachArrayValue(
+  realm: Realm,
+  array: ArrayValue,
+  mapFunc: (element: Value, descriptor: Descriptor) => void
+): void {
+  let lengthValue = Get(realm, array, "length");
+  invariant(lengthValue instanceof NumberValue, "TODO: support non-numeric length on forEachArrayValue");
+  let length = lengthValue.value;
+  for (let i = 0; i < length; i++) {
+    let elementProperty = array.properties.get("" + i);
+    let elementPropertyDescriptor = elementProperty && elementProperty.descriptor;
+    if (elementPropertyDescriptor) {
+      let elementValue = elementPropertyDescriptor.value;
+      if (elementValue instanceof Value) {
+        mapFunc(elementValue, elementPropertyDescriptor);
+      }
+    }
+  }
+}
+
+export function mapArrayValue(
+  realm: Realm,
+  array: ArrayValue,
+  mapFunc: (element: Value, descriptor: Descriptor) => Value
+): ArrayValue {
   let lengthValue = Get(realm, array, "length");
-  invariant(lengthValue instanceof NumberValue, "Invalid length on ArrayValue during reconcilation");
+  invariant(lengthValue instanceof NumberValue, "TODO: support non-numeric length on mapArrayValue");
   let length = lengthValue.value;
+  let newArray = Create.ArrayCreate(realm, length);
+  let returnTheNewArray = false;
+
   for (let i = 0; i < length; i++) {
     let elementProperty = array.properties.get("" + i);
     let elementPropertyDescriptor = elementProperty && elementProperty.descriptor;
-    invariant(elementPropertyDescriptor, `Invalid ArrayValue[${i}] descriptor`);
-    let elementValue = elementPropertyDescriptor.value;
-    if (elementValue instanceof Value) {
-      mapFunc(elementValue, elementPropertyDescriptor);
+    if (elementPropertyDescriptor) {
+      let elementValue = elementPropertyDescriptor.value;
+      if (elementValue instanceof Value) {
+        let newElement = mapFunc(elementValue, elementPropertyDescriptor);
+        if (newElement !== elementValue) {
+          returnTheNewArray = true;
+        }
+        Create.CreateDataPropertyOrThrow(realm, newArray, "" + i, newElement);
+        continue;
+      }
     }
+    Create.CreateDataPropertyOrThrow(realm, newArray, "" + i, realm.intrinsics.undefined);
```

Comment:
This turns missing and accessor elements into undefined elements. Is that really intended?

Reply:
Yep, that's what I intended to do. We're resolving empty slots into undefined.